### ISSUE TITLE
tsc doesn't catch use of unassigned variable inside closure

### ISSUE THREAD

**Issue by steabert** [microsoft/TypeScript#32025]
**TypeScript Version:** 3.6.0-dev.20190621

**Search Terms:**

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
let a: number[]

const b = () => {
  a.push(2) // no 2454 error
}

b() // runtime error

```

**Expected behavior:**

Compile time error 2454.

**Actual behavior:**

No error during compiling with `tsc`, runtime error instead.

**Playground Link:** 

**Related Issues:** 

Not really, i found a bunch of bugs related to use before assign, but not this. It seems this should be a fairly obvious bug, or I'm doing something wrong.

**Comment by nmain** [comment]
Since the error is caught when `b` is replaced by an IIFE, isn't this just https://github.com/microsoft/TypeScript/issues/9998?

**Comment by steabert** [comment]
I'm not sure, but I could see how in this case the compiler would allow it. However, trying to work around it by allowing a to be undefined:
```
let a: number[] | undefined = undefined

a.push(2)
```
made me even more confused as this doesn't get caught even when not in a closure. It seems I'm doing something wrong here.

**Comment by nmain** [comment]
For that one, you need to turn on `strictNullChecks`.

**Comment by steabert** [comment]
Thanks! I had that in my `tsconfig.json` but missed that my test file wasn't included there.

**Comment by ajafff** [comment]
Variables that are never assigned are likely a bug. I can offer a lint rule to detect such variables: https://github.com/fimbullinter/wotan/blob/master/packages/mimir/docs/no-unassigned-variable.md
